README.md
<!-- badges: start -->
  [![R-CMD-check](https://github.com/nrennie/ggflowchart/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/nrennie/ggflowchart/actions/workflows/R-CMD-check.yaml)
  [![codecov](https://codecov.io/gh/nrennie/ggflowchart/branch/main/graph/badge.svg)](https://app.codecov.io/gh/nrennie/ggflowchart?branch=main)
  [![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/ggflowchart)](https://cran.r-project.org/package=ggflowchart)
<!-- badges: end -->

# {ggflowchart} <img src="man/figures/logo.png" align="right" width="120" />

{ggflowchart} is an R package for producing flowcharts using {ggplot2}.

## Installation

Install the package from CRAN using:

```r
install.packages("ggflowchart")
```
or install the development version from GitHub:

```r
remotes::install_github("nrennie/ggflowchart")
```

## Usage

The idea of {ggflowchart} is to create simple flowcharts with minimal effort. Currently, all flowcharts are constructed using the `ggflowchart()` function. For the most basic flowchart, it takes as input a data frame containing (at least) two columns for the start and end points of the edges in the flowchart.

```r
data <- tibble::tibble(from = c("A", "A", "A", "B", "C", "F"),
                       to = c("B", "C", "D", "E", "F", "G"))
```
The flowchart is then created using `ggflowchart()`.

```r
ggflowchart(data)
```

![](man/figures/README-minimal.png)

See [vignettes](https://nrennie.github.io/ggflowchart/articles/) for further examples of usage.

## Aesthetic mappings

Some arguments can be mapped to aesthetics and some cannot. Arguments that can currently be mapped to aesthetics:

* `fill`: must be a valid colour, or the name of a column in `node_data` if mapping to a variable. Fill colours can be changed with e.g. `scale_fill_manual()`.
* `text_colour` (or `text_color`): must be a valid colour, or the name of a column in `node_data` if mapping to a variable.  Colours can be changed with e.g. `scale_colour_manual()`.

Arguments that *may* be mapped to aesthetics in future:

* `colour` (or `color`) for the outline colour of nodes.
* `arrow_colour` (or `arrow_color`)
* `arrow_linewidth`
* `arrow_linetype`
* `alpha`

## Upcoming features

Upcoming features that are currently listed as issues being worked on include:

* Ability to have elliptical (or circular) nodes rather than just rectangles.

If you have a suggestion of an additional feature, or find a bug, please file an issue on the [GitHub repository](https://github.com/nrennie/ggflowchart/issues).

## Contributor guidelines

If you'd like to contribute to {ggflowchart}, I'd welcome your help. If you're making a PR, please follow the guidelines below, to make the collaboration easier:

- You have updated the NEWS and version number in DESCRIPTION.
- You have checked that R CMD check passes with no ERRORs or WARNINGs. If there is a NOTE - please outline what it is in the PR.
- You have checked that `lintr::lint_package()` passes.
- You have checked the list of packages in Imports is still in alphabetical order to enable better tracking of dependencies as the package grows.
- You have not used the base R `|>` pipe (we're not quite ready to specify R 4.1 or higher as a dependency yet!).
- If this is a feature request PR (not a bug fix) please make sure it relates to an issue that has not been assigned to someone else (and tag the issue in the PR description).

If these checks fail, and there is no response from the PR author for 1 month, the PR will be automatically closed.





NEWS.md
## ggflowchart 1.0.0.9000+ 2023_12_20

* Add vignette on how to style nodes
* Add initial unit testing
* Add utils function to determine arrow type
* Add lintr checks to GH Actions
* Add contributor guidelines
* Add option to parse text in nodes.
* Update README
* Update linting
* Add link to issues in Description
* Add `alpha` as argument
* Same level arrows support (issue #5)
* Add option for custom layout (issue #11)
* Allow `x_nudge` and `y_nudge` to vary for each node. (issue #12)
* Add `arrow_linewidth` and `arrow_linewidth` as arguments (issue #14)
* Add `linewidth` as argument (issue #14)
* Add ability to add arrow labels (issue #15)
* Add codecov GitHub action (#24)
* More informative error handling (issue #29)
* Set minimum version of {ggplot2} (issue #31)

## ggflowchart 1.0.0 2023_05_10

* Initial CRAN submission

## ggflowchart 0.0.2 2023_05_10

* Add `horizontal` argument to `ggflowchart()`
* Add `arrow_size` argument to `ggflowchart()`
* Add `text_size` argument to `ggflowchart()`
* Add option to set `fill` based on column in `node_data`
* Add option to set `text_colour` based on column in `node_data`
* Update vignette examples

## ggflowchart 0.0.1 2022_08_06

* Initialise package
* Adds non-exported `get_layout()` to obtain centre of node positions
* Adds non-exported `get_nodes()` to define perimeter of node boxes
* Adds non-exported `add_node_attr()` to include node attributes including labels
* Adds non-exported `get_edges()` to define arrow start and end positions
* Adds `ggflowchart()` to create flowchart


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(ggflowchart)

test_check("ggflowchart")


tests/testthat/test-column-tests.R
test_that("Single column fails", {
  expect_error(ggflowchart(data = data.frame(from = c(1, 2, 3))))
})


R/get_edges.R
#' Get edges
#'
#' Define the end points of the arrows
#'
#' @param data Data frame or tibble of edges. Must have at least two columns, if not labelled "from"
#' and "to" it is assumed first column are "from" node names,
#' second column is "to" node names. Node names must be unique.
#' @param plot_nodes Tibble of node coordinates output from \code{get_nodes()}.
#' @param node_layout Data frame of node layout returned by \code{get_layout()}
#' @importFrom dplyr %>%
#' @importFrom rlang .data
#' @return A tibble.
#' @noRd

get_edges <- function(data, plot_nodes, node_layout) {
  if (ncol(data) < 2) {
    stop("Incorrect number of columns in data: input data should contain at least two columns.")
  }
  if (!all(c("from", "to") %in% colnames(data))) {
    colnames(data)[1:2] <- c("from", "to")
  }
  # define edge types
  edge_type_data <- arrow_direction(
    data = data,
    plot_nodes = plot_nodes,
    node_layout = node_layout
  )
  # compute start and end points for each type of edge
  # Downwards (default)
  plot_edge_down <- edge_type_data %>%
    dplyr::filter(.data$arrow_direction == "down") %>%
    dplyr::select(c(.data$from, .data$to)) %>%
    dplyr::mutate(id = paste("down", dplyr::row_number())) %>%
    tidyr::pivot_longer(
      cols = c("from", "to"),
      names_to = "s_e",
      values_to = "name"
    ) %>%
    dplyr::left_join(
      dplyr::select(plot_nodes, -c(.data$x_nudge, .data$y_nudge)),
      by = "name"
    ) %>%
    dplyr::select(-c(
      .data$y,
      .data$xmin,
      .data$xmax
    )) %>%
    dplyr::mutate(y = ifelse(.data$s_e == "from",
      .data$ymin,
      .data$ymax
    )) %>%
    dplyr::select(-c(
      .data$ymin,
      .data$ymax
    ))
  # Upwards
  plot_edge_up <- edge_type_data %>%
    dplyr::filter(.data$arrow_direction == "up") %>%
    dplyr::select(c(.data$from, .data$to)) %>%
    dplyr::mutate(id = paste("up", dplyr::row_number())) %>%
    tidyr::pivot_longer(
      cols = c("from", "to"),
      names_to = "s_e",
      values_to = "name"
    ) %>%
    dplyr::left_join(
      dplyr::select(plot_nodes, -c(.data$x_nudge, .data$y_nudge)),
      by = "name"
    ) %>%
    dplyr::select(-c(
      .data$y,
      .data$xmin,
      .data$xmax
    )) %>%
    dplyr::mutate(y = ifelse(.data$s_e == "from",
      .data$ymax,
      .data$ymin
    )) %>%
    dplyr::select(-c(
      .data$ymin,
      .data$ymax
    ))
  # Left to right
  plot_edge_lr <- edge_type_data %>%
    dplyr::filter(.data$arrow_direction == "lr") %>%
    dplyr::select(c(.data$from, .data$to)) %>%
    dplyr::mutate(id = paste("lr", dplyr::row_number())) %>%
    tidyr::pivot_longer(
      cols = c("from", "to"),
      names_to = "s_e",
      values_to = "name"
    ) %>%
    dplyr::left_join(
      dplyr::select(
        plot_nodes, -c(.data$x_nudge, .data$y_nudge)
      ),
      by = "name"
    ) %>%
    dplyr::select(-c(
      .data$x,
      .data$ymin,
      .data$ymax
    )) %>%
    dplyr::mutate(x = ifelse(.data$s_e == "from",
      .data$xmax,
      .data$xmin
    )) %>%
    dplyr::select(-c(
      .data$xmin,
      .data$xmax
    ))
  # right to left
  plot_edge_rl <- edge_type_data %>%
    dplyr::filter(.data$arrow_direction == "rl") %>%
    dplyr::select(c(.data$from, .data$to)) %>%
    dplyr::mutate(id = paste("rl", dplyr::row_number())) %>%
    tidyr::pivot_longer(
      cols = c("from", "to"),
      names_to = "s_e",
      values_to = "name"
    ) %>%
    dplyr::left_join(
      dplyr::select(plot_nodes, -c(.data$x_nudge, .data$y_nudge)),
      by = "name"
    ) %>%
    dplyr::select(-c(
      .data$x,
      .data$ymin,
      .data$ymax
    )) %>%
    dplyr::mutate(x = ifelse(.data$s_e == "from",
      .data$xmin,
      .data$xmax
    )) %>%
    dplyr::select(-c(
      .data$xmin,
      .data$xmax
    ))
  # join data set together
  plot_edges <- rbind(
    plot_edge_up, plot_edge_down, plot_edge_rl, plot_edge_lr
  )
  return(plot_edges)
}


R/get_nodes.R
#' Get nodes
#'
#' Define the corners of the nodes.
#'
#' @param node_layout Data frame of node layout returned by \code{get_layout()}
#' @param x_nudge Distance from centre of edge of node box in x direction.
#' Default 0.35.
#' @param y_nudge Distance from centre of edge of node box in y direction.
#' Default 0.25.
#' @importFrom dplyr %>%
#' @importFrom rlang .data
#' @return A tibble.
#' @noRd

get_nodes <- function(node_layout,
                      x_nudge = 0.35,
                      y_nudge = 0.25) {
  # if x_nudge not specified use global values
  if (!("x_nudge" %in% colnames(node_layout))) {
    node_layout$x_nudge <- x_nudge
  }
  # if y_nudge not specified use global values
  if (!("y_nudge" %in% colnames(node_layout))) {
    node_layout$y_nudge <- y_nudge
  }
  # define corners
  plot_nodes <- node_layout %>%
    dplyr::mutate(xmin = .data$x - .data$x_nudge,
                  xmax = .data$x + .data$x_nudge,
                  ymin = .data$y - .data$y_nudge,
                  ymax = .data$y + .data$y_nudge)
  return(plot_nodes)
}


R/get_layout.R
#' Get layout
#'
#' Obtains the coordinates for the nodes of the flowchart from the input edges
#'
#' @param data Data frame or tibble of edges. Must have at least two columns, if not labelled "from"
#' it is assumed first column is `"from"` node names,
#' second column is `"to"` node names. Node names must be unique.
#' @param layout One of `c("tree", "custom")`. If `"tree"` uses the tree layout
#' from {igraph}. If `"custom"`, then `x` and `y` columns must be provided in
#' `node_data` specifying the coordinates of the centre of the boxes. Default
#' `"tree"`.
#' @importFrom dplyr %>%
#' @return A tibble.
#' @noRd

get_layout <- function(data, layout = "tree", node_data = NULL) {
  # check number of columns
  if (ncol(data) < 2) {
    stop("Incorrect number of columns in data: data should contain at least two columns.")
  }
  # define notin
  "%notin%" <- function(x, y) {
    !("%in%"(x, y))
  }
  # check column names
  if ("from" %in% colnames(data) && "to" %notin%  colnames(data)) {
    stop("data contains only 'from' column but not 'to'")
  }
  if ("to" %in% colnames(data) && "from" %notin%  colnames(data)) {
    stop("data contains only 'to' column but not 'from'")
  }
  if (!all(c("from", "to") %in% colnames(data))) {
    colnames(data)[1:2] <- c("from", "to")
    message("'from' and 'to' not found in column names, using first two columns instead.")
  }
  # generate layout
  if (layout == "tree") {
    data <- dplyr::select(data, c(.data$from, .data$to))
    g <- igraph::graph_from_data_frame(data, directed = TRUE)
    coords <- igraph::layout_as_tree(g)
    colnames(coords) <- c("x", "y")
    output <- tibble::as_tibble(coords) %>%
      dplyr::mutate(name = igraph::vertex_attr(g, "name"))
  } else if (layout == "custom") {
    # check if names correct if any missing
    if ("name" %notin% colnames(node_data)) {
      stop("node_data must have a column called `name` to join to node layout")
    }
    # check if x and y exist
    if (any(c("x", "y") %notin% colnames(node_data))) {
      stop("node_data must have a columns called `x` and `y` with node coordinates")
    }
    output <- node_data %>%
      dplyr::select(c(.data$x, .data$y, .data$name))
  }
  return(output)
}


R/add_node_attr.R
#' Add node attributes
#'
#' Adds additional information to the node e.g. variable to colour by
#'
#' @param node_layout Data frame of node layout returned by \code{get_layout()}.
#' @param node_data Data frame or tibble of node information. Must have at least one column
#' called "name" for node names to join by. To change labels that appear on nodes, include
#' a column with the name "label".
#' @return A tibble.
#' @noRd

add_node_attr <- function(node_layout, node_data) {
  # define notin
  "%notin%" <- function(x, y) {
    !("%in%" (x, y))
  }
  # if no additional node info provided
  if (is.null(node_data)) {
    return(node_layout)
  } else {
    # check if names correct if any missing
    if ("name" %notin% colnames(node_data)) {
      stop("node_data must have a column called `name` to join to node layout")
    }
    # add node attributes
    node_layout <- dplyr::left_join(node_layout, node_data, by = "name")
    return(node_layout)
  }
}


R/arrow_direction.R
#' Single arrow direction
#'
#' Define the direction of a single arrow
#'
#' @param plot_nodes Tibble of node coordinates output from \code{get_nodes()}.
#' @param from Character name of starting node.
#' @param to Character name of end node.
#' @importFrom rlang .data
#' @return A tibble.
#' @noRd

single_arrow_direction <- function(plot_nodes,
                                   from,
                                   to) {
  # calculate slope
  from_coords <- unlist(dplyr::filter(plot_nodes, .data$name == from)[, c("x", "y")])
  to_coords <- unlist(dplyr::filter(plot_nodes, .data$name == to)[, c("x", "y")])
  x_diff <- to_coords["x"] - from_coords["x"]
  y_diff <- (to_coords["y"] - from_coords["y"])
  # arrow type
  if (y_diff == 0 && x_diff > 0) {
    return("lr")
  } else if (y_diff == 0 && x_diff < 0) {
    return("rl")
  } else if (y_diff > 0) {
    return("up")
  } else {
    return("down")
  }
}

#' Arrow direction
#'
#' Define the direction of each arrow
#' @param data Data frame or tibble of edges. Must have two columns, first
#' column are "from" node names, second column is "to" node names.
#' Node names must be unique.
#' @param plot_nodes Tibble of node coordinates output from \code{get_nodes()}.
#' @param node_layout Data frame of node layout returned by \code{get_layout()}
#' @importFrom rlang .data
#' @return A tibble.
#' @noRd

arrow_direction <- function(data,
                            plot_nodes,
                            node_layout) {
  data %>%
    dplyr::mutate(
      arrow_direction = purrr::map2(
        .x = data$from,
        .y = data$to,
        .f = ~ single_arrow_direction(node_layout, .x, .y)
      ) %>%
        unlist()
    )
}


R/get_edge_labels.R
#' Get edge labels
#'
#' Define the end points of the arrows
#'
#' @param data Data frame or tibble of edges. Must have at least two columns, if not labelled "from"
#' and "to" it is assumed first column are "from" node names,
#' second column is "to" node names. Node names must be unique.
#' @param plot_edges Tibble of edge coordinates output from \code{get_edges()}.
#' @importFrom dplyr %>%
#' @importFrom rlang .data
#' @return A tibble.
#' @noRd

get_edge_labels <- function(
    data,
    plot_edges) {
  if (is.null(data[["label"]])) {
    return(NULL)
  } else {
    # check number of columns
    if (ncol(data) < 2) {
      stop("Incorrect number of columns in data: data should contain at least two columns.")
    }
    # define notin
    "%notin%" <- function(x, y) {
      !("%in%"(x, y))
    }
    # check column names
    if ("from" %in% colnames(data) && "to" %notin%  colnames(data)) {
      stop("data contains only 'from' column but not 'to'")
    }
    if ("to" %in% colnames(data) && "from" %notin%  colnames(data)) {
      stop("data contains only 'to' column but not 'from'")
    }
    if (!all(c("from", "to") %in% colnames(data))) {
      colnames(data)[1:2] <- c("from", "to")
      message("'from' and 'to' not found in column names, using first two columns instead.")
    }
    edge_labels <- plot_edges %>%
      dplyr::group_by(.data$id) %>%
      dplyr::mutate(
        text_x = mean(.data$x),
        text_y = mean(.data$y)
      ) %>%
      dplyr::ungroup() %>%
      dplyr::select(.data$s_e, .data$name, .data$text_x, .data$text_y) %>%
      tidyr::pivot_wider(
        names_from = .data$s_e,
        values_from = .data$name
      ) %>%
      dplyr::left_join(
        data,
        by = c("from", "to")
      ) %>%
      dplyr::select(.data$text_x, .data$text_y, .data$label)
    return(edge_labels)
  }
}


R/ggflowchart.R
#' Generate a flowchart in ggplot2
#'
#' Generates the flowchart
#'
#' @param data Data frame or tibble of edges. Must have two columns, first
#' column are `"from"` node names, second column is `"to"` node names. Node names must be unique.
#' @param node_data Data frame or tibble of node information. If not `NULL`,
#' must have at least one column called "name" for node names to join by.
#' Default `NULL`.
#' @param layout One of `c("tree", "custom")`. If `"tree"` uses the tree layout
#' from {igraph}. If `"custom"`, then `x` and `y` columns must be provided in
#' `node_data` specifying the coordinates of the centre of the boxes. Default
#' `"tree"`.
#' @param fill Fill colour of nodes. Must be a valid colour name or hex
#' code, or the name of a column in node_data (quoted or unquoted).
#' Column names take priority over names of colours. Default `"white"`.
#' @param colour Outline colour of nodes. Must be a valid colour name or hex
#' code. Default `"black"`.
#' @param linewidth Width of node outlines. Default 0.5.
#' @param alpha Transparency of fill colour in nodes. Default 1.
#' @param text_colour Colour of labels in nodes. Must be a valid colour name
#' or hex code, or the name of a column in node_data (quoted or unquoted).
#' Column names take priority over names of colours. Default `"black"`.
#' @param text_size Font size of labels in nodes. Default 3.88.
#' @param family Font family for node labels. Default `"sans"`.
#' @param parse If TRUE, the labels will be parsed into expressions
#' and displayed as described in ?plotmath. Default `FALSE`.
#' @param arrow_colour Colour of arrows. Must be a valid colour name or hex
#' code. Default `"black"`.
#' @param arrow_size Size of arrow head. Default 0.3.
#' @param arrow_linewidth Linewidth of arrow lines. Default 0.5.
#' @param arrow_linetype Linetype of arrow lines. Default `"solid"`.
#' @param arrow_label_fill Fill colour of arrow labels. Default `"white"`.
#' @param x_nudge Distance from centre of edge of node box in x direction.
#' Ignored if `x_nudge` is a column in `node_data`. Default 0.35.
#' @param y_nudge Distance from centre of edge of node box in y direction.
#'  Ignored if `y_nudge` is a column in `node_data`. Default 0.25.
#' @param horizontal Boolean specifying if flowchart should go from left to
#' right. Default `FALSE`.
#' @param color Outline colour of nodes - overrides colour. Must be a valid colour name or hex
#' code. Default `NULL`.
#' @param text_color Colour of labels in nodes - overrides text_colour. Must be a valid colour name
#' or hex code. Default `NULL`.
#' @param arrow_color Colour of arrows - overrides arrow_colour. Must be a valid colour name or hex
#' code. Default `NULL`.
#' @importFrom rlang .data
#' @return A ggplot2 object.
#' @export
#' @examples
#' data <- tibble::tibble(from = c("A", "A", "A", "B", "C", "F"), to = c("B", "C", "D", "E", "F", "G"))
#' ggflowchart(data)
ggflowchart <- function(data,
                        node_data = NULL,
                        layout = "tree",
                        fill = "white",
                        colour = "black",
                        linewidth = 0.5,
                        alpha = 1,
                        text_colour = "black",
                        text_size = 3.88,
                        family = "sans",
                        parse = FALSE,
                        arrow_colour = "black",
                        arrow_size = 0.3,
                        arrow_linewidth = 0.5,
                        arrow_linetype = "solid",
                        arrow_label_fill = "white",
                        x_nudge = 0.35,
                        y_nudge = 0.25,
                        horizontal = FALSE,
                        color = NULL,
                        text_color = NULL,
                        arrow_color = NULL) {
  # define notin
  "%notin%" <- function(x, y) {
    !("%in%"(x, y))
  }
  # check layout valid
  if (layout %notin% c("tree", "custom")) {
    stop('Layout must be one of c("tree", "custom").')
  }
  # check data has non-zero rows
  if (nrow(data) < 1) {
    stop("data must have at least one row.")
  }
  # convert arguments
  fill <- rlang::ensym(fill)
  text_colour <- rlang::ensym(text_colour)
  # set colours
  if (!is.null(color)) {
    colour <- color
  }
  if (!is.null(rlang::enexpr(text_color))) {
    text_colour <- rlang::ensym(text_color)
  }
  if (!is.null(arrow_color)) {
    arrow_colour <- arrow_color
  }
  # define position of nodes
  node_layout <- get_layout(
    data = data,
    layout = layout,
    node_data = node_data
  )
  # add edge attributes
  if (layout == "custom") {
    node_layout <- add_node_attr(
      node_layout = node_layout,
      node_data = dplyr::select(node_data, -dplyr::any_of(c("x", "y")))
    )
  } else {
    node_layout <- add_node_attr(
      node_layout = node_layout,
      node_data = node_data
    )
  }
  # define edges of node rectangles
  plot_nodes <- get_nodes(
    node_layout = node_layout,
    x_nudge = x_nudge,
    y_nudge = y_nudge
  )
  # check if labels exist as a column,
  # if not, add it as a duplicate of name
  if ("label" %notin% colnames(plot_nodes)) {
    plot_nodes <- dplyr::mutate(plot_nodes, label = .data$name)
  }
  # define arrows
  plot_edges <- get_edges(
    data = data,
    plot_nodes = plot_nodes,
    node_layout = node_layout
  )
  # define edge labels
  edge_labels <- get_edge_labels(
    data = data,
    plot_edges = plot_edges
  )
  # create the flowchart
  p <- ggplot2::ggplot()
  # add nodes
  if (as.character(fill) %in% colnames(node_data)) {
    p <- p +
      ggplot2::geom_rect(
        data = plot_nodes,
        mapping = ggplot2::aes(
          xmin = .data$xmin,
          ymin = .data$ymin,
          xmax = .data$xmax,
          ymax = .data$ymax,
          fill = !!fill
        ),
        alpha = alpha,
        colour = colour,
        linewidth = linewidth
      )
  } else {
    p <- p +
      ggplot2::geom_rect(
        data = plot_nodes,
        mapping = ggplot2::aes(
          xmin = .data$xmin,
          ymin = .data$ymin,
          xmax = .data$xmax,
          ymax = .data$ymax
        ),
        alpha = alpha,
        colour = colour,
        fill = as.character(fill),
        linewidth = linewidth
      )
  }
  # add text
  if (as.character(text_colour) %in% colnames(node_data)) {
    p <- p +
      ggplot2::geom_text(
        data = plot_nodes,
        mapping = ggplot2::aes(
          x = .data$x,
          y = .data$y,
          label = .data$label,
          colour = !!text_colour
        ),
        family = family,
        size = text_size,
        parse = parse
      )
  } else {
    p <- p +
      ggplot2::geom_text(
        data = plot_nodes,
        mapping = ggplot2::aes(
          x = .data$x,
          y = .data$y,
          label = .data$label
        ),
        family = family,
        size = text_size,
        parse = parse,
        colour = as.character(text_colour)
      )
  }
  # add arrows
  p <- p +
    ggplot2::geom_path(
      data = plot_edges,
      mapping = ggplot2::aes(
        x = .data$x,
        y = .data$y,
        group = .data$id
      ),
      arrow = ggplot2::arrow(
        length = ggplot2::unit(arrow_size, "cm"),
        type = "closed"
      ),
      linewidth = arrow_linewidth,
      linetype = arrow_linetype,
      colour = arrow_colour
    ) +
    ggplot2::theme_void() +
    ggplot2::theme(
      plot.margin = ggplot2::unit(c(0.5, 0.5, 0.5, 0.5), unit = "cm")
    )
  # add arrow edge labels
  if (!is.null(edge_labels)) {
    p <- p +
      ggplot2::geom_label(
        data = edge_labels,
        mapping = ggplot2::aes(
          x = .data$text_x,
          y = .data$text_y,
          label = .data$label
        ),
        fill = arrow_label_fill,
        family = family,
        size = text_size
      )
  }
  # check if horizontal
  if (horizontal == TRUE) {
    p <- p +
      ggplot2::coord_flip() +
      ggplot2::scale_y_reverse()
  }
  return(p)
}
